plenum/server/consensus/consensus_data_provider.py

```diff
@@ -48,22 +60,85 @@ def is_primary(self) -> bool:
     def preprepared(self) -> List[PrePrepare]:
         """
         List of PrePrepare messages, for which quorum of Prepare messages is not reached yet
-        TODO: Make it simple field?
         """
-        return []
+        return self._preprepared
 
     @property
     def prepared(self) -> List[PrePrepare]:
         """
         List of PrePrepare messages, for which quorum of Prepare messages is reached
-        TODO: Make it simple field?
         """
-        return []
+        return self._prepared
 
     @property
     def stable_checkpoint(self) -> int:
-        return 0
+        return self._stable_checkpoint
 
     @property
     def checkpoints(self) -> List[Checkpoint]:
-        return []
+        """
+        List of Checkpoints, which has been sent, but not stabilized yet
+        """
+        return self._checkpoints.values()
+
+    def preprepare_batch(self, pp: PrePrepare):
+        """
+        After pp had validated, it placed into _preprepared list
+        """
+        self._preprepared.append(pp)
```

Comment:
There is quite a lot of checks in other similar functions, but not this one. For example - what if preprepare was already in preprepared or prepared list? What if preprepare is below already existing checkpoint? Should we do these checks here or somewhere upper in the stack? If latter we'd better document expected preconditions for this function.

Reply:
comment will be left to move checks upper. additional checks added